orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -788,6 +788,35 @@ IMPLEMENT_GRADIENT_BUILDER(GetReduceMeanGradient) {
   return result;
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetReduceLogSumExpGradient) {
+  std::vector<NodeDef> result;
+  auto attributes = SrcNodeAttributes();
+  bool keepdims = true;
+  if (attributes.find("keepdims") != attributes.end() &&
+      attributes.at("keepdims").has_i()) {
+    keepdims = static_cast<bool>(attributes.at("keepdims").i());
+  }
+
+  ArgDef grad = GO(0);
+  if (!keepdims && attributes.find("axes") != attributes.end()) {
+    std::vector<int64_t> axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+    grad = IA("Unsqueezed_Grad");
+    result.push_back(NodeDef("Unsqueeze", {GO(0)}, {grad}, {MakeAttribute("axes", axes_values)}));
```

Comment:
I think neither Unsqueeze are needed here, since Sub and Mul both supports broadcasting.

Reply:
I thought about it, but since this op takes a dims as input essentially users can choose which dim to reduce, we need to reply on which axes to unsqueeze based on the input dims? I just wanted to avoid corner cases since bcast only aligns either the leftmost or the rightmost.

Reply:
yes, you are right. 
think of the case [A, B, C] reduced to [A, C] with axes = [1] and keepdim=0, there are actuality not broadcast-able.